### PR TITLE
Extract State Machine Logic from SegmentBuilder (#97)

### LINKED ISSUE #97: Extract State Machine Logic from SegmentBuilder
## Parent Issue

Part of #91 - Refactor SegmentBuilder state management to prevent recurring bugs

## Overview

Extract all state transition logic from SegmentBuilder into pure functions that define a complete, explicit state machine. This eliminates implicit state dependencies and makes all transitions testable.

## Current State

State transitions are scattered across 10+ handler functions:
- State reset pattern duplicated 3 times (lines 341-345, 393-417, 477-491)
- No single source of truth for valid transitions
- Easy to forget to update all 8 state variables
- useEffect auto-advance creates hidden transitions (lines 141-157)

## Goals

1. Extract all state transitions to pure functions
2. Create complete, explicit state machine
3. Achieve 100% test coverage on transitions
4. Make state machine self-documenting through types

## Dependencies

**Requires:**
- Issue #95 - Extract Validation & Types Module (for `SegmentBuilderState`, `Step` types)
- Issue #96 - Extract Segment Building Logic (for segment building functions)

### Step 1: Create File Structure

Create new file:
```
frontend/src/components/routes/SegmentBuilder/
  transitions.ts              # NEW: State transition functions
  transitions.test.ts         # NEW: Unit tests
```

### Step 2: Implement State Transition Functions

**File: `transitions.ts`**

```typescript
import { StationResponse, LineResponse } from '@/lib/api'
import { SegmentBuilderState, Step } from './types'

/**
 * Transitions to initial state (select station)
 * 
 * Use when: starting new route, canceling, resetting
 * 
 * @returns Complete initial state
 */
export function transitionToSelectStation(): SegmentBuilderState {
  return {
    currentStation: null,
    selectedLine: null,
    nextStation: null,
    step: 'select-station',
    error: null,
  }
}

/**
 * Transitions to select line step
 * 
 * Use when: station selected that has multiple lines
 * 
 * @param currentStation - Station that was selected
 * @returns Complete state for line selection
 */
export function transitionToSelectLine(
  currentStation: StationResponse
): SegmentBuilderState {
  return {
    currentStation,
    selectedLine: null,
    nextStation: null,
    step: 'select-line',
    error: null,
  }
}

/**
 * Transitions to select next station step
 * 
 * Use when: line selected (or auto-selected if only one line)
 * 
 * @param currentStation - Current station
 * @param selectedLine - Line to travel on
 * @returns Complete state for next station selection
 */
export function transitionToSelectNextStation(
  currentStation: StationResponse,
  selectedLine: LineResponse
): SegmentBuilderState {
  return {
    currentStation,
    selectedLine,
    nextStation: null,
    step: 'select-next-station',
    error: null,
  }
}

/**
 * Transitions to choose action step
 * 
 * Use when: next station selected, now user chooses continue or mark as destination
 * 
 * @param currentStation - Current station
 * @param selectedLine - Current line
 * @param nextStation - Next station selected
 * @returns Complete state for action choice
 */
export function transitionToChooseAction(
  currentStation: StationResponse,
  selectedLine: LineResponse,
  nextStation: StationResponse
): SegmentBuilderState {
  return {
    currentStation,
    selectedLine,
    nextStation,
    step: 'choose-action',
    error: null,
  }
}

/**
 * Transitions to error state (preserves current state, adds error)
 * 
 * Use when: validation fails or API error occurs
 * 
 * @param currentState - Current state
 * @param error - Error message to display
 * @returns State with error message
 */
export function transitionWithError(
  currentState: SegmentBuilderState,
  error: string
): SegmentBuilderState {
  return {
    ...currentState,
    error,
  }
}

/**
 * Clears error from current state
 * 
 * Use when: user takes action after error
 * 
 * @param currentState - Current state with possible error
 * @returns State with error cleared
 */
export function transitionClearError(
  currentState: SegmentBuilderState
): SegmentBuilderState {
  return {
    ...currentState,
    error: null,
  }
}

/**
 * Computes next state after segments are updated
 * 
 * Use when: segments change (continue, destination, delete, edit)
 * 
 * This implements the "auto-advance" logic currently in useEffect
 * 
 * @param currentState - Current state
 * @param hasSegments - Whether route has any segments
 * @returns Next state based on segment presence
 */
export function computeStateAfterSegmentUpdate(
  currentState: SegmentBuilderState,
  hasSegments: boolean
): SegmentBuilderState {
  // If we have segments and we're at choose-action, stay there
  if (currentState.step === 'choose-action' && hasSegments) {
    return currentState
  }
  
  // If we have segments and current/selected are set, transition to next station
  if (
    hasSegments &&
    currentState.currentStation &&
    currentState.selectedLine
  ) {
    return transitionToSelectNextStation(
      currentState.currentStation,
      currentState.selectedLine
    )
  }
  
  // Otherwise, reset to initial state
  return transitionToSelectStation()
}

/**
 * Computes state for resuming from existing segments
 * 
 * Use when: editing route, after deletion
 * 
 * @param resumeStation - Station to resume from (from computeResumeState)
 * @param resumeLine - Line to resume on (from computeResumeState)
 * @returns State ready to continue building
 */
export function transitionToResumeFromSegments(
  resumeStation: StationResponse | null,
  resumeLine: LineResponse | null
): SegmentBuilderState {
  if (!resumeStation || !resumeLine) {
    return transitionToSelectStation()
  }
  
  return transitionToSelectNextStation(resumeStation, resumeLine)
}

/**
 * Validates state transition is allowed
 * 
 * Use for: debugging, invariant checking
 * 
 * @param from - Current step
 * @param to - Target step
 * @returns true if transition is valid
 */
export function isValidTransition(from: Step, to: Step): boolean {
  const validTransitions: Record<Step, Step[]> = {
    'select-station': ['select-line', 'select-next-station'],
    'select-line': ['select-next-station'],
    'select-next-station': ['choose-action'],
    'choose-action': ['select-next-station', 'select-station'],
  }
  
  return validTransitions[from].includes(to)
}
```

### Step 3: Comprehensive Unit Tests

**File: `transitions.test.ts`**

Test coverage requirements:
- All transition functions: 100% statement, 100% branch coverage
- Invariant checking: all fields set correctly
- Edge cases: null handling, state preservation

Example test structure:
```typescript
import { describe, it, expect } from 'vitest'
import {
  transitionToSelectStation,
  transitionToSelectLine,
  transitionToSelectNextStation,
  transitionToChooseAction,
  transitionWithError,
  transitionClearError,
  computeStateAfterSegmentUpdate,
  transitionToResumeFromSegments,
  isValidTransition
} from './transitions'

describe('transitionToSelectStation', () => {
  it('should return complete initial state', () => {
    const state = transitionToSelectStation()
    
    expect(state.currentStation).toBeNull()
    expect(state.selectedLine).toBeNull()
    expect(state.nextStation).toBeNull()
    expect(state.step).toBe('select-station')
    expect(state.error).toBeNull()
  })
  
  it('should clear all previous state', () => {
    // Verify it doesn't preserve anything from previous state
    const state = transitionToSelectStation()
    expect(Object.keys(state)).toEqual([
      'currentStation',
      'selectedLine',
      'nextStation',
      'step',
      'error'
    ])
  })
})

describe('transitionToSelectLine', () => {
  const mockStation = { tfl_id: '123', name: 'Test Station' } as StationResponse
  
  it('should set correct state for line selection', () => {
    const state = transitionToSelectLine(mockStation)
    
    expect(state.currentStation).toBe(mockStation)
    expect(state.selectedLine).toBeNull()
    expect(state.nextStation).toBeNull()
    expect(state.step).toBe('select-line')
    expect(state.error).toBeNull()
  })
})

describe('transitionToSelectNextStation', () => {
  const mockStation = { tfl_id: '123', name: 'Test Station' } as StationResponse
  const mockLine = { tfl_id: 'northern', name: 'Northern' } as LineResponse
  
  it('should set correct state for next station selection', () => {
    const state = transitionToSelectNextStation(mockStation, mockLine)
    
    expect(state.currentStation).toBe(mockStation)
    expect(state.selectedLine).toBe(mockLine)
    expect(state.nextStation).toBeNull()
    expect(state.step).toBe('select-next-station')
    expect(state.error).toBeNull()
  })
})

describe('transitionToChooseAction', () => {
  const mockStation1 = { tfl_id: '123', name: 'Station 1' } as StationResponse
  const mockStation2 = { tfl_id: '456', name: 'Station 2' } as StationResponse
  const mockLine = { tfl_id: 'northern', name: 'Northern' } as LineResponse
  
  it('should set correct state for action choice', () => {
    const state = transitionToChooseAction(mockStation1, mockLine, mockStation2)
    
    expect(state.currentStation).toBe(mockStation1)
    expect(state.selectedLine).toBe(mockLine)
    expect(state.nextStation).toBe(mockStation2)
    expect(state.step).toBe('choose-action')
    expect(state.error).toBeNull()
  })
})

describe('transitionWithError', () => {
  it('should preserve state and add error', () => {
    const currentState = transitionToSelectStation()
    const state = transitionWithError(currentState, 'Test error')
    
    expect(state.currentStation).toBeNull()
    expect(state.selectedLine).toBeNull()
    expect(state.nextStation).toBeNull()
    expect(state.step).toBe('select-station')
    expect(state.error).toBe('Test error')
  })
  
  it('should preserve complex state', () => {
    const mockStation = { tfl_id: '123', name: 'Test' } as StationResponse
    const mockLine = { tfl_id: 'northern', name: 'Northern' } as LineResponse
    const currentState = transitionToSelectNextStation(mockStation, mockLine)
    
    const state = transitionWithError(currentState, 'Error message')
    
    expect(state.currentStation).toBe(mockStation)
    expect(state.selectedLine).toBe(mockLine)
    expect(state.error).toBe('Error message')
  })
})

describe('transitionClearError', () => {
  it('should clear error while preserving state', () => {
    const stateWithError = {
      currentStation: { tfl_id: '123' } as StationResponse,
      selectedLine: null,
      nextStation: null,
      step: 'select-station' as Step,
      error: 'Some error'
    }
    
    const state = transitionClearError(stateWithError)
    
    expect(state.error).toBeNull()
    expect(state.currentStation).toBe(stateWithError.currentStation)
    expect(state.step).toBe('select-station')
  })
})

describe('computeStateAfterSegmentUpdate', () => {
  it('should stay at choose-action if has segments', () => {
    const currentState = transitionToChooseAction(
      { tfl_id: '123' } as StationResponse,
      { tfl_id: 'northern' } as LineResponse,
      { tfl_id: '456' } as StationResponse
    )
    
    const state = computeStateAfterSegmentUpdate(currentState, true)
    
    expect(state.step).toBe('choose-action')
  })
  
  it('should transition to select-next-station if has segments and station/line set', () => {
    const mockStation = { tfl_id: '123' } as StationResponse
    const mockLine = { tfl_id: 'northern' } as LineResponse
    const currentState = {
      currentStation: mockStation,
      selectedLine: mockLine,
      nextStation: null,
      step: 'select-line' as Step,
      error: null
    }
    
    const state = computeStateAfterSegmentUpdate(currentState, true)
    
    expect(state.step).toBe('select-next-station')
    expect(state.currentStation).toBe(mockStation)
    expect(state.selectedLine).toBe(mockLine)
  })
  
  it('should reset to select-station if no segments', () => {
    const currentState = transitionToChooseAction(
      { tfl_id: '123' } as StationResponse,
      { tfl_id: 'northern' } as LineResponse,
      { tfl_id: '456' } as StationResponse
    )
    
    const state = computeStateAfterSegmentUpdate(currentState, false)
    
    expect(state.step).toBe('select-station')
    expect(state.currentStation).toBeNull()
  })
})

describe('transitionToResumeFromSegments', () => {
  it('should transition to select-next-station when station and line provided', () => {
    const mockStation = { tfl_id: '123' } as StationResponse
    const mockLine = { tfl_id: 'northern' } as LineResponse
    
    const state = transitionToResumeFromSegments(mockStation, mockLine)
    
    expect(state.step).toBe('select-next-station')
    expect(state.currentStation).toBe(mockStation)
    expect(state.selectedLine).toBe(mockLine)
  })
  
  it('should transition to select-station when no station', () => {
    const state = transitionToResumeFromSegments(null, { tfl_id: 'northern' } as LineResponse)
    
    expect(state.step).toBe('select-station')
  })
  
  it('should transition to select-station when no line', () => {
    const state = transitionToResumeFromSegments({ tfl_id: '123' } as StationResponse, null)
    
    expect(state.step).toBe('select-station')
  })
})

describe('isValidTransition', () => {
  it('should allow valid transitions', () => {
    expect(isValidTransition('select-station', 'select-line')).toBe(true)
    expect(isValidTransition('select-station', 'select-next-station')).toBe(true)
    expect(isValidTransition('select-line', 'select-next-station')).toBe(true)
    expect(isValidTransition('select-next-station', 'choose-action')).toBe(true)
    expect(isValidTransition('choose-action', 'select-next-station')).toBe(true)
    expect(isValidTransition('choose-action', 'select-station')).toBe(true)
  })
  
  it('should reject invalid transitions', () => {
    expect(isValidTransition('select-line', 'select-station')).toBe(false)
    expect(isValidTransition('select-next-station', 'select-line')).toBe(false)
    expect(isValidTransition('select-line', 'choose-action')).toBe(false)
  })
})
```

### Step 4: Add Transition Logging (Optional but Recommended)

**File: `transitions.ts` (add helper)**

```typescript
/**
 * Logs state transition (useful for debugging)
 * 
 * @param from - Previous state
 * @param to - New state
 * @param action - Description of action that caused transition
 */
export function logTransition(
  from: SegmentBuilderState,
  to: SegmentBuilderState,
  action: string
): void {
  if (process.env.NODE_ENV === 'development') {
    console.log('[SegmentBuilder Transition]', {
      action,
      from: {
        step: from.step,
        currentStation: from.currentStation?.name,
        selectedLine: from.selectedLine?.name,
        nextStation: from.nextStation?.name,
        hasError: !!from.error
      },
      to: {
        step: to.step,
        currentStation: to.currentStation?.name,
        selectedLine: to.selectedLine?.name,
        nextStation: to.nextStation?.name,
        hasError: !!to.error
      }
    })
  }
}
```

### Step 5: Refactor SegmentBuilder to Use Transitions

Update SegmentBuilder.tsx to import and use transition functions:

**Before** (lines 341-345):
```typescript
setCurrentStation(null)
setSelectedLine(null)
setNextStation(null)
setStep('select-station')
setError(null)
```

**After**:
```typescript
const newState = transitionToSelectStation()
setState(newState)
// Or apply individual setters if not using single state object yet
```

Apply this pattern to all state transitions.

### Unit Tests (NEW)
- [ ] All transition functions: 20-25 test cases
  - Each function tested with valid inputs
  - State shape verified (all 5 fields present)
  - Null handling
  - Edge cases

- [ ] `isValidTransition`: 10-12 test cases
  - All valid transitions
  - All invalid transitions
  - Boundary cases

### Integration Tests (EXISTING)
- [ ] All 22+ existing SegmentBuilder tests must pass
- [ ] No behavior changes (transitions identical, just moved)

### Coverage Goals
- [ ] Transition functions: 100% statement, 100% branch coverage
- [ ] SegmentBuilder.tsx: No decrease in coverage

## Acceptance Criteria

- [ ] `transitions.ts` created with all state transition functions
- [ ] `transitions.test.ts` created with 100% coverage
- [ ] All functions have JSDoc comments
- [ ] Optional logging helper implemented
- [ ] SegmentBuilder.tsx refactored to use transitions (no scattered state updates)
- [ ] All 22+ existing tests pass
- [ ] No behavior changes (only refactoring)
- [ ] Pre-commit hooks pass
- [ ] TypeScript strict mode compliant
- [ ] ESLint compliant

## File Locations

**New files:**
- `frontend/src/components/routes/SegmentBuilder/transitions.ts`
- `frontend/src/components/routes/SegmentBuilder/transitions.test.ts`

**Modified files:**
- `frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx`

## Estimated Effort

- Create transitions.ts: 2-3 hours
- Create transitions.test.ts: 3-4 hours
- Refactor SegmentBuilder.tsx: 1-2 hours
- Testing & verification: 0.5-1 hour

**Total: 6-8 hours**

## Blocked By

- #95 - Extract Validation & Types Module (for types)
- #96 - Extract Segment Building Logic (for segment functions)

## Blocks

- #97 - Integrate Pure Functions (needs all modules)

## Success Metrics

- **Before**: State transitions scattered across 10+ handlers, 3 duplicated reset patterns
- **After**: Single source of truth for all transitions, 100% test coverage, 0 duplication

## References

- Parent issue: #91
- Required: #95, #96
- Original SegmentBuilder: `frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx`

---

**Status:** Ready for Implementation (after #95, #96 complete)
**Assignee:** TBD (can be executed by any agent)

### PR DESCRIPTION
**Pull request by mnbf9rca**
## Summary
Extracts all state transition logic from SegmentBuilder component into pure, testable functions as part of Phase 3 refactoring (#91). This eliminates 62 scattered state setter calls and replaces them with 12 explicit transition functions.

### Core State Machine (transitions.ts - NEW)
- **12 transition functions** returning complete `CoreSegmentBuilderState` objects
- **Pure functions**: same input → same output, no side effects
- **Complete state**: each function returns all 5 core state fields
- **Explicit transitions**: no hidden state dependencies

Key functions:
- `transitionToSelectStation()` - Reset to initial state
- `transitionToSelectLine()` - Advance to line selection
- `transitionToSelectNextStation()` - Advance to next station selection
- `transitionToChooseAction()` - Advance to action choice
- `transitionBackFromChooseAction()` - Navigate back one step
- `transitionToResumeFromSegments()` - Resume from deletion/edit
- `computeStateAfterStationSelect()` - Auto-advance logic based on line count
- `computeStateAfterSegmentUpdate()` - Handle segment changes
- `transitionWithError()` / `transitionClearError()` - Error handling
- `isValidTransition()` / `logTransition()` - Debugging utilities

### Test Coverage (transitions.test.ts - NEW)
- **42 tests** covering all transition functions
- **95.83% coverage** on new code
- Tests verify state shape, null handling, edge cases

### Component Refactoring (SegmentBuilder.tsx)
- **All 11 handlers** now use transition functions
- **Zero duplication** of reset patterns (was 3 occurrences)
- **Added `applyTransition()` helper** to update component state from transition results
- **Removed obsolete functions**: `resumeFromLastSegment`, `computeResumeState` calls

### Type System (types.ts)
- **Added `CoreSegmentBuilderState`** - 5-field interface for pure transitions
- **Separation of concerns**: Core state machine vs UI state (isSaving, saveSuccess)

### Scope Expansion - Truncation Context Fix (segments.ts)
**Problem**: When deleting segment B from route A→B→C→D, user should resume from station A, not reset to initial state

**Solution**: Enhanced `deleteSegmentAndResequence()` to return truncation context
- **Added `DeletionResult` interface** with segments + resumeFrom fields
- **Function signature updated** to accept stations and lines parameters
- **Returns both segments and resume context** (station/line from truncation point)

### Updated Tests (segments.test.ts)
- **Updated 22 existing tests** to expect DeletionResult
- **Added 5 new truncation tests** covering edge cases
- All 35 segment tests passing

## Test Results
```
✓ All 362 tests passing (no regressions)
✓ 42 new transition tests (95.83% coverage)
✓ 5 new truncation tests
✓ Pre-commit hooks passing (prettier, eslint, typescript)
```

## Architecture
Follows ADR 11: Frontend State Management Pattern - Pure Function Architecture

**State machine flow:**
```
select-station ←→ select-line → select-next-station → choose-action
     ↑                                                        ↓
     └────────────────────────────────────────────────────────
```

## Related Issues
- Closes #97
- Part of #91 (Phase 3: Pure Functions Architecture)
- Builds on #95 (Extract Validation & Types) - merged
- Builds on #96 (Extract Segment Building Logic) - merged

## Migration Path
No breaking changes - component API unchanged. Internal state management refactored to use pure functions.

## Summary by Sourcery

Extract and centralize SegmentBuilder state logic into pure, testable transition functions, enhance segment deletion to provide resume context, and refactor the component to use these transitions while updating types and tests accordingly.

New Features:
- Extract core state machine transitions into pure functions in transitions.ts
- Introduce DeletionResult with resumeFrom context in deleteSegmentAndResequence

Bug Fixes:
- Fix truncation context so deleting segments resumes from correct station and line instead of resetting state

Enhancements:
- Refactor SegmentBuilder component to use applyTransition helper and pure transition functions, removing inline state setters and obsolete helpers
- Separate core state machine type (CoreSegmentBuilderState) from UI-specific state

Tests:
- Add 42 unit tests covering all transition functions (95.83% coverage)
- Update 22 existing segment tests and add 5 truncation context tests for deleteSegmentAndResequence

### CODE REVIEW

File: frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx [hunk 1]
```diff
@@ -343,43 +346,16 @@ export function SegmentBuilder({
...
   const handleBackFromChooseAction = () => {
-    setNextStation(null)
-    setStep('select-next-station')
-    setError(null)
+    if (!currentStation || !selectedLine) return
+
+    const newState = transitionBackFromChooseAction(currentStation, selectedLine)
+    applyTransition(newState)
     // Don't exit edit mode - keep saveSuccess as false so we stay in editing state
   }
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx:371]: No changes made

 Rationale:
 - All other handlers use silent guard clauses for null checks
 - Guard clauses represent defensive programming for impossible states
 - Adding logging here creates inconsistency unless added to ALL handlers
 - Consider as separate broader refactoring if needed
 
 ok?

File: frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx [hunk 2]
```diff
@@ -182,23 +191,21 @@ export function SegmentBuilder({
...
-    setError(null)
+    if (!currentStation) return
+    const newState = transitionToSelectNextStation(currentStation, line)
+    applyTransition(newState)
   }
 
   const handleNextStationSelect = (stationId: string | undefined) => {
-    if (!stationId) {
+    if (!stationId || !currentStation || !selectedLine) {
       setNextStation(null)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx]: ACCEPT WITH MODIFICATION

 File: frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx (line 200)

 Change: Replace setNextStation(null) with transitionBackFromChooseAction() when:
 - User clears the combobox (!stationId)
 - Required state is missing (!currentStation || !selectedLine)

 Also add explicit handling for station lookup failure with console warning.

 Rationale:
 - Aligns with state machine architecture from Issue #97
 - More consistent with handleStationSelect which uses transition functions
 - Fixes two issues: inconsistent state handling + silent station lookup failure

File: frontend/src/components/routes/SegmentBuilder/segments.ts [hunk 3]
```diff
@@ -20,6 +20,21 @@ import type { SegmentRequest } from './types'
...
+ */
+export interface DeletionResult {
+  /** Remaining segments after deletion (resequenced) */
+  segments: SegmentRequest[]
+  /** Context for resuming route building after truncation */
+  resumeFrom: {
+    /** Station to resume from (from last remaining segment) */
+    station: StationResponse | null
+    /** Line to resume on (from last remaining segment) */
+    line: LineResponse | null
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply frontend/src/components/routes/SegmentBuilder/segments.ts]: ACCEPT
 File: frontend/src/components/routes/SegmentBuilder/segments.ts (lines 26-36)

 Change: Enhance JSDoc to document when resumeFrom.station and resumeFrom.line are null
 - Add explanation that both fields are null when deleting from sequence 0 or lookup fails
 - Clarify that null values trigger transition back to initial state

 Rationale: Low risk, high value for maintainability. The implementation already handles null, but documentation doesn't explain when it occurs.
